src/win32/fs.cpp

```diff
@@ -1408,102 +1424,211 @@ void WinDirNotify::process(DWORD dwBytes)
         }
     }
 #endif
+    clientWaiter->notify();
 }
 
 // request change notifications on the subtree under hDirectory
 void WinDirNotify::readchanges()
 {
+    assert( std::this_thread::get_id() == smNotifierThread->get_id());
+
 #ifndef WINDOWS_PHONE
-    if (ReadDirectoryChangesW(hDirectory, (LPVOID)notifybuf[active].data(),
-                              (DWORD)notifybuf[active].size(), TRUE,
+    if (notifybuf.size() != 65534)
+    {
+        notifybuf.resize(65534);
+    }
+    auto readRet = ReadDirectoryChangesW(hDirectory, (LPVOID)notifybuf.data(),
+                              (DWORD)notifybuf.size(), TRUE,
                               FILE_NOTIFY_CHANGE_FILE_NAME
                             | FILE_NOTIFY_CHANGE_DIR_NAME
                             | FILE_NOTIFY_CHANGE_LAST_WRITE
                             | FILE_NOTIFY_CHANGE_SIZE
                             | FILE_NOTIFY_CHANGE_CREATION,
-                              &dwBytes, &overlapped, completion))
+                              &dwBytes, &overlapped, completion);
+
+    if (!readRet && GetLastError() == ERROR_INVALID_PARAMETER)
+    {
+        // ReadDirectoryChangesW fails with ERROR_INVALID_PARAMETER when the buffer length is greater than 64 KB and the application is monitoring a directory over the network. This is due to a packet size limitation with the underlying file sharing protocols.
```

Comment:
We used to use two pre-allocated buffers that were exactly 65534 bytes long - that comment is text from the WIN32 documentation indicating why we don't use more.  Initially I was looking at increasing the size of the buffer to make sure we had enough space, but as it turned out no amount was enough (even testing with 4MB), because we always needed more on the next batch.  So for a while I did have some code in there to reduce the size back down to 64KB if we got that error.    But now with the threading solution, we only use up to 4KB in my testing (on a very fast machine).   I think retaining the 65534 size is sensible for now, it's already half the size we used to use since we are only allocating one of those buffers now, and usually we will just copy out 4KB at then reuse that buffer.   If it gets larger than that then we'll swap the buffer for processing, and allocate 64KB again for the next batch.     I've moved the comment and clarified it, as it's useful to know why we chose 65534, I think.  thanks

Reply:
Actually, good point, we don't need the second call anymore since ERROR_INVALID_PARAMETER  won't occur since we are calling with just under 64KB.   I've removed that block too, thanks!